File: Realize/8Week/PracticeTrie.py
class Node:
    def __init__(self):
        self.child = {}
        self.is_terminal = False

class Trie:
    def __init__(self):
        self.root = Node()

    def insert(self, word: str):
        cur = self.root
        for i, c in enumerate(word):
            if c not in cur.child:
                cur.child[c] = Node()
            cur = cur.child[c]
            if i == len(word) - 1:
                cur.is_terminal = True

    def search(self, word: str) -> bool:
        cur = self.root
        for i, c in enumerate(word):
            if c not in cur.child:
                return False
            cur = cur.child[c]
            if i == len(word) - 1 and not cur.is_terminal:
                return False
        return True

    def delete(self, word: str):
        deleted = self._delete(self.root, word, 0)
        if deleted:
            print(f"{word} 삭제 완료")
        else:
            print(f"{word} 단어가 없습니다.")

    def _delete(self, node: Node, word: str, idx: int) -> bool:
        c = word[idx]
        if c not in node.child:
            return False

        cur = node.child[c]
        idx += 1

        if idx == len(word):
            if not cur.is_terminal:
                return False
            cur.is_terminal = False
            if not cur.child:
                del node.child[c]
        else:
            if not self._delete(cur, word, idx):
                return False
            if not cur.is_terminal and not cur.child:
                del node.child[c]
        return True


# Test code
if __name__ == "__main__":
    trie = Trie()
    trie.insert("apple")
    trie.insert("april")
    trie.insert("app")
    trie.insert("ace")
    trie.insert("bear")
    trie.insert("best")

    print(trie.search("apple"))   # True
    print(trie.search("april"))   # True
    print(trie.search("app"))     # True
    print(trie.search("ace"))     # True
    print(trie.search("bear"))    # True
    print(trie.search("best"))    # True
    print(trie.search("abc"))     # False

    print()
    trie.delete("apple")
    print(trie.search("apple"))   # False
    print(trie.search("april"))   # True
    print(trie.search("appl"))    # False
    trie.delete("apple")          # Already deleted

File: Realize/PracticeTrie.cpp
#include <iostream>
#include <unordered_map>
using namespace std;

class Node {
public:
    unordered_map<char, Node*> child;
    bool isTerminal;

    Node() {
        isTerminal = false;
    }
};

class Trie {
public:
    Node* root;

    Trie() {
        root = new Node();
    }

    void insert(const string& str) {
        Node* cur = root;
        for (int i = 0; i < str.length(); i++) {
            char c = str[i];
            if (cur->child.find(c) == cur->child.end()) {
                cur->child[c] = new Node();
            }
            cur = cur->child[c];

            if (i == str.length() - 1) {
                cur->isTerminal = true;
                return;
            }
        }
    }

    bool search(const string& str) {
        Node* cur = root;
        for (int i = 0; i < str.length(); i++) {
            char c = str[i];
            if (cur->child.find(c) == cur->child.end()) {
                return false;
            }
            cur = cur->child[c];

            if (i == str.length() - 1 && !cur->isTerminal) {
                return false;
            }
        }
        return true;
    }

    bool deleteHelper(Node* node, const string& str, int idx) {
        char c = str[idx];
        if (node->child.find(c) == node->child.end()) {
            return false;
        }

        Node* cur = node->child[c];
        idx++;

        if (idx == str.length()) {
            if (!cur->isTerminal) return false;
            cur->isTerminal = false;
            if (cur->child.empty()) {
                delete cur;
                node->child.erase(c);
            }
        }
        else {
            if (!deleteHelper(cur, str, idx)) return false;
            if (!cur->isTerminal && cur->child.empty()) {
                delete cur;
                node->child.erase(c);
            }
        }
        return true;
    }

    void deleteWord(const string& str) {
        bool result = deleteHelper(root, str, 0);
        if (result) {
            cout << str << " 삭제 완료" << endl;
        }
        else {
            cout << str << " 단어가 없습니다." << endl;
        }
    }
};

int main() {
    Trie trie;
    trie.insert("apple");
    trie.insert("april");
    trie.insert("app");
    trie.insert("ace");
    trie.insert("bear");
    trie.insert("best");

    cout << boolalpha;
    cout << trie.search("apple") << endl;  // true
    cout << trie.search("april") << endl;  // true
    cout << trie.search("app") << endl;    // true
    cout << trie.search("ace") << endl;    // true
    cout << trie.search("bear") << endl;   // true
    cout << trie.search("best") << endl;   // true
    cout << trie.search("abc") << endl;    // false

    cout << endl;
    trie.deleteWord("apple");
    cout << trie.search("apple") << endl;  // false
    cout << trie.search("april") << endl;  // true
    cout << trie.search("appl") << endl;   // false
    trie.deleteWord("apple");              // already deleted
}
